02a-dbc.md

```diff
@@ -0,0 +1,153 @@
+>
+> *   Design by Contract is a way of using Assertions for interface specification.
+> *   Design by Contract defines pre-, post- and invariant conditions a function and a caller agree to obey.
+
+In [Design by Contract](https://en.wikipedia.org/wiki/Design_by_contract), the interaction between a caller
+and a library is treated as being governed by a *contract*. A contract involves three different types of requirements.
+
+* Pre-conditions: Things that must be true before calling the function.
+* Post-conditions: Things that are guaranteed to be true after the function returns.
+* Invariant-conditions: Things that are guaranteed not to change during the function's execution
+
+**Note**: In Python's contracts, only pre- and post-conditions are handled.
+
+In the examples here, we use [PyContracts](https://andreacensi.github.io/contracts/index.html#) which use
+Python [decorator](https://www.python.org/dev/peps/pep-0318) notation. In addition, to simplify the examples,
+the following imports are assumed...
+
+~~~ {.python}
+from math import sqrt
+from contracts import contract, new_contract
+~~~
+
+To demonstrate the use of contracts, in the example here, we implement our own version of an integer square root
+function for perfect squares, called psqrt. We define a contract that indicates the caller is required to pass an
+integer value greater than or equal to zero. This is an example of a pre-condition. Next, the function is required
+to return an integer greater than or equal to zero. This is an example of a post-condition.
+
+~~~ {.python}
+@contract(x='int,>=0',returns='int,>=0')
+def psqrt(x):
+    retval = sqrt(x)
+    iretval = int(retval)
+    return iretval if iretval == retval else retval
+~~~
+
+Now, lets see what happens when we use this function to compute square roots.
+
+~~~ {.output}
+>>> pqsrt(4)
+2
+>>> psqrt(81)
+9
+~~~
+
+Values of 4 and 81 are both integers. So, in these cases the caller has obeyed the pre-conditions of the contract.
+In addition, because both 4 and 81 are perfect squares, the function correctly returns their integer square root.
+So, the funtion has obeyd the post-conditions of the contract.
+
+Now, lets see what happens when the caller failes to obey the pre-conditions of the contract by passing a negative
+number.
+
+~~~ {.output}
+>>> psqrt(-4)
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+  File "<decorator-gen-2>", line 2, in psqrt
+  File "/Library/Python/2.7/site-packages/PyContracts/contracts/main.py", line 253, in contracts_checker
+    raise e
+contracts.interface.ContractNotRespected: Breach for argument 'x' to psqrt().
+Condition -4 >= 0 not respected
+checking: >=0       for value: Instance of <type 'int'>: -4   
+checking: int,>=0   for value: Instance of <type 'int'>: -4   
+Variables bound in inner context:
+~~~
+
+An exception is raised indicating a failure to obey the pre-condition for passing a value greather than or equal to zero.
```

Comment:
at this point, the students haven't seen exceptions. That is in the next part of the lesson (3).